retry.go

```diff
@@ -75,18 +80,58 @@ type retry http.Client
 // If sending a request to one endpoint fails due to e.g. a network
 // or DNS error, Send tries the next endpoint. It aborts once the
 // context is canceled or its deadline exceeded.
-func (r *retry) Send(ctx context.Context, method string, endpoints []string, path string, body io.ReadSeeker, options ...requestOption) (*http.Response, error) {
+//
+// Any endpoint that fails to respond gets marked offline for some
+// time period. Offline endpoints will be marked online periodically.
+func (lb *loadBalancer) Send(ctx context.Context, client *retry, method string, endpoints []string, path string, body io.ReadSeeker, options ...requestOption) (*http.Response, error) {
 	if len(endpoints) == 0 {
 		return nil, errors.New("kes: no server endpoint")
 	}
+	if len(endpoints) == 1 {
+		request, err := http.NewRequestWithContext(ctx, method, endpoint(endpoints[0], path), retryBody(body))
+		if err != nil {
+			return nil, err
+		}
+		for _, opt := range options {
+			opt(request)
+		}
+		response, err := client.Do(request)
+		if errors.Is(err, context.Canceled) {
+			return nil, err
+		}
+		if errors.Is(err, context.DeadlineExceeded) {
+			return nil, err
+		}
+		if urlErr, ok := err.(*url.Error); ok {
+			if connErr, ok := urlErr.Err.(*ConnError); ok {
+				return nil, connErr
+			}
+		}
+		return response, err
+	}
+
 	var (
 		request  *http.Request
 		response *http.Response
 		err      error
 		R        = rand.Intn(len(endpoints)) // randomize endpoints => avoid hitting the same endpoint all the time.
 	)
+
+retry:
 	for i := range endpoints {
 		nextEndpoint := endpoints[(i+R)%len(endpoints)]
+
+		lb.lock.Lock()
+		t, ok := lb.endpoints[nextEndpoint]
+		if ok && !t.IsZero() && t.Before(time.Now().Add(5*time.Minute)) {
+			lb.lock.Unlock()
+			continue
+		}
+		if !ok || !t.IsZero() {
+			lb.endpoints[nextEndpoint] = time.Time{}
+		}
```

Comment:
Could we change this slightly?

We set it to current time, but defer a call that will clear it if the call goes fine?

This means that only one attempt will be made on a previously bad node. If this call goes fine we resume using it for all calls.

This is just to reduce the number of calls made until we have confirmed it is fine.

Reply:
Oh - I see now. A 2nd concurrent request could try the same node since it has been marked as online already. Yeah, that we can change 👍

Reply:
@vadmeste If the loop falls through with no attempts made, it will reset all connections. 

@aead - please correct me if I am wrong.

Reply:
> I think one goroutine doing healthcheck in background is probably fine and easier

@vadmeste I explicitly wanted to avoid a background job that needs to be monitored and/or cancelled.